### PR TITLE
Fixes #12722 and #12723 - extending capsule show page with pulp content

### PR DESCRIPTION
**Pull request by tstrachota**
- extends capsule content api with actions to get sync status and
  cancelling synchronizations
- adds panel with sync progress
    - buttons for starting and cancelling synchronizations
    - progressbar for currently running synchroniations, links to proxy
      related foreman-tasks
    - displays last synchronization in case it failed
- adds tab with repository content details
    - repositories split by content view
    - grouped by lifecycle environment

**TODO:**
- [x] date and yes/no formatting in templates
- [x] tests for the UI js part
- [x] tests for the backend part

Please test with the latest Foreman develop.

### CODE REVIEW

File: app/lib/katello/capsule_content.rb [hunk 1]
```diff
@@ -53,6 +53,49 @@ def consumer
       @consumer
     end
 
+    def active_sync_tasks
+      ForemanTasks::Task.for_resource(self.capsule).where(:result => 'pending')
```
Comment (ohadlevy) [inline app/lib/katello/capsule_content.rb]: nitpicky:  ForemanTasks::Task.for_resource(self.capsule) DRY

File: app/lib/katello/capsule_content.rb [hunk 2]
```diff
@@ -53,6 +53,49 @@ def consumer
...
+      ForemanTasks::Task.for_resource(self.capsule).where('started_at > ?', last_sync_time).where.not(:result => 'pending')
+    end
+
+    def last_sync_time
+      task = ForemanTasks::Task.for_resource(self.capsule).where.not(:ended_at => nil).where(:result => 'success').order(:ended_at).last
+      task.ended_at unless task.nil?
+    end
+
+    def environment_syncable?(env)
+      !env.content_view_environments.where('updated_at > ?', last_sync_time).empty?
```
Comment (ohadlevy) [inline app/lib/katello/capsule_content.rb]: nitpick: be positive and use ```.any?``` :)

File: app/lib/katello/capsule_content.rb [hunk 3]
```diff
@@ -53,6 +53,49 @@ def consumer
...
+        @pulp_repositories
+      end
+    end
+
+    def cancel_sync
+      tasks = active_sync_tasks
+      tasks.each do |t|
+        t.cancel
+      end
+      tasks
```
Comment (ohadlevy) [inline app/lib/katello/capsule_content.rb]: nitpick: the entire method can be written as 
```ruby
def cancel_sync
  active_sync_tasks.map(&:cancel)
end

File: app/lib/katello/util/thread_session.rb [hunk 4]
```diff
@@ -54,9 +54,16 @@ def self.current=(o)
             end
 
             def self.pulp_config(user_remote_id, &_block)
-              uri = URI.parse(SETTINGS[:katello][:pulp][:url])
+              Katello.pulp_server = runcible_instance(SETTINGS[:katello][:pulp][:url], user_remote_id)
+              yield if block_given?
```
Comment (ohadlevy) [inline app/lib/katello/util/thread_session.rb]: :+1: TBH not sure why pulp server needs to be in a thread? anyway /cc @shlomizadok as you had similar issue

File: app/services/katello/pulp/content_counts_calculator.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+          :puppet_repositories => 0,
+          :errata => 0
+        }
+
+        @repos.each do |r|
+          if is_repo_type?(r, 'rpm')
+            counts[:yum_repositories] += 1
+            counts[:packages] += get_unit_count(r, 'rpm')
+            counts[:errata] += get_unit_count(r, 'erratum')
+          elsif is_repo_type?(r, 'docker')
```
Comment (ohadlevy) [inline app/services/katello/pulp/content_counts_calculator.rb]: nitpick: use case?

File: app/services/katello/pulp/content_counts_calculator.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+      protected
+
+      def is_repo_type?(repo, repo_type)
+        repo['notes'] && (repo['notes']['_repo-type'] == "#{repo_type}-repo")
+      end
+
+      def get_unit_count(repo, unit_type)
+        if repo['content_unit_counts']
+          repo['content_unit_counts'][unit_type] || 0
```
Comment (ohadlevy) [inline app/services/katello/pulp/content_counts_calculator.rb]: should it be 0 or N/A ?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        };
+
+        $scope.syncStatusText = function (syncState, syncStatus) {
+            if (syncStatus === undefined)
+                return;
+
+            if (syncState.is(syncState.SYNCING)) {
+                return translate("Capsule currently syncing to your locations...");
+            } else if (syncState.is(syncState.SYNC_TRIGGERED)) {
+                return translate("Starting the synchronization...");
```
Comment (ohadlevy) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: not an English speaker, but i think it should be:
```Synchronization Started (or is about to start)```

and the same for Cancel
```Synchronization is now cancelled (or being cancelled).```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                                .replace('%count', syncableEnvs.length)
+                                .replace('%envs', envNames);
+                } else {
+                    return translate("Capsule is synchronized");
+                }
+            }
+        };
+
+        $scope.toggleExpandEnvironment = function (environment) {
+            $scope.expandEnvironments[environment.id] = !$scope.expandEnvironments[environment.id];
```
Comment (ohadlevy) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: nitpick: missing return

File: app/controllers/katello/api/v2/capsule_content_controller.rb [hunk 9]
```diff
@@ -54,6 +54,23 @@ def sync
...
+    api :DELETE, '/capsules/:id/content/sync',  'Cancel running capsule synchronization.'
+    param :id, Integer, :desc => 'Id of the capsule', :required => true
+    def cancel_sync
+      tasks = capsule_content.cancel_sync
+      if tasks.empty?
+        render_message _('There\'s no running synchronization for this capsule.')
+      else
+        render_message _('Trying to cancel the synchronization...')
+      end
+    end
```
Comment (ehelms) [inline app/controllers/katello/api/v2/capsule_content_controller.rb]: @jlsherrill can you please look this over in case there is some conflict with the Pulp nodes removal work?

File: app/controllers/katello/api/v2/capsule_content_controller.rb [hunk 10]
```diff
@@ -54,6 +54,23 @@ def sync
       respond_for_async :resource => task
     end
 
+    api :GET, '/capsules/:id/content/sync',  'Synchronize the content to the capsule'
```
Comment (ehelms) [inline app/controllers/katello/api/v2/capsule_content_controller.rb]: Description doesn't match the intent of the action and is missing i18n

File: app/models/katello/glue/pulp/node.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
+module Katello
+  module Glue::Pulp::Node
```
Comment (ehelms) [inline app/models/katello/glue/pulp/node.rb]: There is a team working on removing the Pulp node concept and replacing it with direct calls to the Pulp on a capsule, you may want to coordinate this with them @jlsherrill @johnpmitsch @cfouant

File: app/services/katello/pulp/content_counts_calculator.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+      def calculate
+        counts = {
+          :packages => 0,
+          :puppet_modules => 0,
+          :yum_repositories => 0,
+          :docker_repositories => 0,
+          :puppet_repositories => 0,
+          :errata => 0
+        }
```
Comment (ehelms) [inline app/services/katello/pulp/content_counts_calculator.rb]: What about package groups?

File: app/services/katello/pulp/content_counts_calculator.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+        @repos.each do |r|
```
Comment (ehelms) [inline app/services/katello/pulp/content_counts_calculator.rb]: Single letter variables aren't  very readable to me.

File: app/views/foreman/smart_proxies/_content_tab.html.erb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      <th class=""><%= s_("Content View Name") %></th>
+      <th class=""><%= s_("Composite") %></th>
+      <th class=""><%= s_("Last Published") %></th>
+      <th class=""><%= s_("Hosts") %></th>
+      <th class=""><%= s_("Products") %></th>
+      <th class=""><%= s_("Yum repos") %></th>
+      <th class=""><%= s_("Docker repos") %></th>
+      <th class=""><%= s_("Packages") %></th>
+      <th class=""><%= s_("Errata") %></th>
+      <th class=""><%= s_("Puppet modules") %></th>
```
Comment (ehelms) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: What is `s_` here? We typically use the translate directive, e.g.

```
<th translate>Puppet Modules</th>
```

File: app/views/foreman/smart_proxies/_content_tab.html.erb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    </tr>
+  </thead>
+  <tbody>
+    <tr ng-repeat-start="env in syncStatus.lifecycle_environments">
+      <td colspan="10"
+        ng-click="toggleExpandEnvironment(env)"
+        class="expander {{isEnvronmentExpanded(env) ? '' : 'collapsed'}}">
```
Comment (ehelms) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: Do you mind lining up the attributes? That makes it a bit easier to read:

```
<td colspan="10"
      ng-click=
      class=
```

File: app/views/foreman/smart_proxies/_content_tab.html.erb [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        <span class="caret"></span>
+        {{ env.name }}
+      </td>
+    </tr>
+    <tr ng-repeat-end ng-repeat="cv in env.content_views" ng-show="isEnvronmentExpanded(env)">
+      <td>
+        <a href="/content_views/{{cv.id}}/versions" target="_self">
```
Comment (ehelms) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: This could be written using states instead (https://github.com/Katello/katello/blob/master/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/content-views.routes.js#L70):

```
ui-sref="content-views.versions({versionId: cv.id)"
```

File: app/views/foreman/smart_proxies/_content_tab.html.erb [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+          {{cv.name}}
```
Comment (ehelms) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: `{{ cv.name }}` for consistency

File: app/views/katello/api/v2/capsule_content/sync_status.json.rabl [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+          :content_hosts => content_view.systems.readable.count,
+          :products => content_view.products.enabled.count
+        }
+      }
+      repo_data = @capsule_content.pulp_repositories_data(env, content_view)
+      attributes[:counts].merge!(Katello::Pulp::ContentCountsCalculator.new(repo_data).calculate)
+      attributes
+    end
+  end
+end
```
Comment (ehelms) [inline app/views/katello/api/v2/capsule_content/sync_status.json.rabl]: I'll check on this in testing as well, but I feel like this could create a rather large JSON output.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+/**
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js]: I feel like this could simply live under capsules/. What do you think?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.routes.js [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+ * @name Bastion.capsule-content.config
+ *
+ * @requires $stateProvider
+ *
+ * @description
+ *   Used for systems level configuration such as setting up the ui state machine.
+ */
+angular.module('Bastion.capsule-content').config(['$stateProvider', function ($stateProvider) {
+
+    // Fake state to prevent redirects
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.routes.js]: Sounds like this is not being handled properly in the main Bastion code, can you describe what happens? I might be able to look into it and fix it to avoid this.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 21]
```diff
@@ -0,0 +1,167 @@
+/**
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Something to consider, there is a lot of code in here that is 'logic' code that lives outside of the need to dictate what the view is doing and would be better off inside a service. You could think in some ways of a service as similar to a Rails helper, or a service object -- it holds business logic (ideally re-usable but we don't always know what we will need to re-use) that is not tied to a specific controller allowing the controller to be thin.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+ *
+ * @description
+ *   Provides the functionality for the capsule-content page.
+ */
+angular.module('Bastion.capsule-content').controller('CapsuleContentController',
+    ['$scope', '$urlMatcherFactory', '$location', 'translate', 'CapsuleContent', 'AggregateTask',
+    function ($scope, $urlMatcherFactory, $location, translate, CapsuleContent, AggregateTask) {
+
+        var urlMatcher = $urlMatcherFactory.compile("/smart_proxies/:capsuleId");
+        var capsuleId = urlMatcher.exec($location.path()).capsuleId;
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: I am assuming this is due to not using states and thus can't rely on `$stateParams` ?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+            },
+            is: function () {
+                return _.contains(arguments, this.state || this.DEFAULT);
+            }
+        }
+
+        function refreshSyncStatus() {
+            CapsuleContent.sync_status({id: capsuleId}, function (syncStatus) {
+                $scope.syncStatus = syncStatus;
+                if (syncStatus.active_sync_tasks.length > 0) {
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: This will cause you linting errors due to requiring camelCase for JS. You can get around it via `syncStatus['active_sync_tasks']`

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+            return (task && (task.state === 'pending' || task.state === 'running'))
+        };
+
+        function stateFromTask(syncTask) {
+            if (isTaskInProgress(syncTask))
+                return syncState.SYNCING;
+            if (syncTask && syncTask.result !== 'success')
+                return syncState.FAILURE;
+            else
+                return syncState.DEFAULT;
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: The linter will probably indicate, but please use `{}` notation for if-else statements.

Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: We tend to prefer a single return statement at the end of functions, for example:

```
function stateFromTask(syncTask) {
    var state;

    if () {
         state = syncState.SYNCING;
   } else {
        state = syncState.FAILURE;
  }

   return state;
}
```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+        };
+
+
+        $scope.syncState = syncState;
+
+        $scope.expandEnvironments = {};
+
+        refreshSyncStatus();
+
+        $scope.$on('$destroy', function () {
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Is there a function of ours generating this? or is this being generated outside of the code in this controller?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+            if (isTaskInProgress(syncTask))
+                return syncState.SYNCING;
+            if (syncTask && syncTask.result !== 'success')
+                return syncState.FAILURE;
+            else
+                return syncState.DEFAULT;
+        };
+
+
+        $scope.syncState = syncState;
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Good indicator of service object when you are creating objects and functions and attaching them to the scope separately. You could re-factor that out to a service 'Capsule' or 'CapsuleSync' and do:

```
     $scope.syncState = CapsuleSync.syncstate;
```

or, just reference CapsuleSync.syncState where you need it.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+
+        $scope.cancelSync = function () {
+            if ($scope.syncState.is(syncState.SYNCING)) {
+
+                $scope.syncState.set(syncState.CANCEL_TRIGGERED);
+                CapsuleContent.cancel_sync({id: capsuleId}, function (response) {
+                }, function (response) {
+                    $scope.syncErrorMessages = [response.data.displayMessage];
+                });
+            }
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: You can avoid an empty success function here by using the promise `$q` API directly:

```
var promise;

promise = CapsuleContent.cancel_sync({id: capsuleId}).$promise;

promise.catch(function (response) {
    $scope.syncErrorMessages = [response.data.displayMessage];
});
```

or

```
CapsuleContent.cancelSync({id: capsuleId}).$promise.callback(function (response) {

});
```

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+        };
+
+        $scope.syncStatusText = function (syncState, syncStatus) {
+            if (syncStatus === undefined)
+                return;
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Please use brackets for if statements. As well, you can use `angular.isUndefined(syncStatus)`

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+                var syncableEnvs = _.where(syncStatus.lifecycle_environments, {syncable: true})
+                if (syncableEnvs.length > 0) {
+                    var envNames = _.pluck(syncableEnvs, 'name').join(', ');
+                    return translate("%count environments can be synchronized: %envs")
+                                .replace('%count', syncableEnvs.length)
+                                .replace('%envs', envNames);
+                } else {
+                    return translate("Capsule is synchronized");
+                }
+            }
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: A single return statement in this function would be nice.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+
+        $scope.isEnvronmentExpanded = function (environment) {
+            return $scope.expandEnvironments[environment.id];
+        };
+
+        $scope.progressbarType = function (syncTask) {
+            if (syncTask === undefined || syncTask.result === 'pending' || syncTask.result === 'success')
+                return 'success';
+            else
+                return 'danger';
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Single return and brackets for the if-else

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.factory.js [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+ * @description
+ *   Provides a BastionResource for capsule content.
+ */
+angular.module('Bastion.capsule-content').factory('CapsuleContent',
+    ['BastionResource', function (BastionResource) {
+
+        return BastionResource('/katello/api/capsules/:id/content/:action', {id: '@id'}, {
+          sync_status: {method: 'GET', isArray: false, params: {action: 'sync'}},
+          sync: {method: 'post', isArray: false, params: {action: 'sync'}},
+          cancel_sync: {method: 'delete', isArray: false, params: {action: 'sync'}}
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.factory.js]: Switch the snake_case to camel_case or the linter will ding you. This is because most JavaScript is by nature camelCase oriented so we chose to write our code this way. API data from the server comes in snake_case, so we use [''] syntax to refrence it and find it helps to identify server vs. code data.

File: app/controllers/katello/api/v2/capsule_content_controller.rb (hunk 9 above)
Comment (jlsherrill) [inline app/controllers/katello/api/v2/capsule_content_controller.rb]: I don't believe so, as long as the cancel_sync function still works.  I will add a user story to ensure that.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js (hunk 22 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Exactly

File: app/services/katello/pulp/content_counts_calculator.rb (hunk 6 above)
Comment (tstrachota) [inline app/services/katello/pulp/content_counts_calculator.rb]: 0 is imho more suitable here because we calculate sum for a content view across repos of multiple types

File: app/views/foreman/smart_proxies/_content_tab.html.erb (hunk 14 above)
Comment (tstrachota) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: `s_` does namespaced translations. Can be just `_` here as well. I can change it to `translate` if that's preferred.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.routes.js (hunk 20 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.routes.js]: Without state routing ends up in router's `otherwise` https://github.com/Katello/bastion/blob/master/app/assets/javascripts/bastion/bastion.module.js#L60-L63 The function does some legacy url transformation I guess. In our case it assigns the same url to `location.url` which causes the reload.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js (hunk 25 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: I thought it's triggered by angular before a scope is destroyed. But TBH I just copied it from https://github.com/Katello/katello/blob/master/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/details/content-view-versions.controller.js#L54-L60

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js (hunk 19 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js]: No problem, I can move it there.

File: app/views/foreman/smart_proxies/_content_tab.html.erb (hunk 16 above)
Comment (tstrachota) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: @ehelms I tried `ui-sref="content-views.details.versions({contentViewId: cv.id})"` It only changes url but UI remains the same. No idea why, the error console is empty. The same problem is with katello menu items (which is worse). Foreman menu links work ok. Do you have a clue?

Comment (ehelms) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: If you inspect the html, does the anchor link have a proper href populated with what you'd expect?

Comment (tstrachota) [inline app/views/foreman/smart_proxies/_content_tab.html.erb]: It does. When I click on it the browser location is set properly to `/content_views/1/versions` but it has no effect on the page content.

File: app/models/katello/glue/pulp/node.rb [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+      base.class_eval do
+        lazy_accessor :pulp_repositories, :initializer => lambda { |_s| pulp_node.extensions.repository.retrieve_all }
+      end
+    end
+
+    module InstanceMethods
+
+      def pulp_node
+        @pulp_node ||= User.runcible_instance(pulp_url, User.remote_user)
+      end
```
Comment (jlsherrill) [inline app/models/katello/glue/pulp/node.rb]: I don't think this will work today with the authentication we use for capsules because of the oauth being different than the pulp server installed on the satellite itself.  There is some work going on as part of the pulp 2.8 work to support using cert based auth to talk to the capsules.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js (hunk 25 above)
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: Ahh ok, I see what you are going for now.

File: app/models/katello/glue/pulp/node.rb (hunk 11 above)
Comment (jlsherrill) [inline app/models/katello/glue/pulp/node.rb]: yeah a lot of this overlaps with the node removal work that is ongoing that will add cert auth to the capsules.  Would you want to meet to discuss this more?

Comment (tstrachota) [inline app/models/katello/glue/pulp/node.rb]: @jlsherrill yes please, let's meet. Ping me whenever you have time.

File: app/models/katello/glue/pulp/node.rb (hunk 32 above)
Comment (tstrachota) [inline app/models/katello/glue/pulp/node.rb]: I thought we are passing the same oauth secret to the capsule-installer.

File: app/models/katello/concerns/smart_proxy_extensions.rb [hunk 33]
```diff
@@ -7,6 +7,10 @@ module SmartProxyExtensions
       PULP_NODE_FEATURE = "Pulp Node"
 
       included do
+        include ForemanTasks::Concerns::ActionSubject
+        include Katello::Glue::Pulp::Node if SETTINGS[:katello][:use_pulp]
+        include Katello::Glue if SETTINGS[:katello][:use_pulp]
```
Comment (ohadlevy) [inline app/models/katello/concerns/smart_proxy_extensions.rb]: nitpick, its the same condition, might simply wrap it in an if
```ruby
if SETTINGS[:katello][:use_pulp]
  include Katello::Glue::Pulp::Node
  include Katello::Glue
end
```

Also, I do not know if the order matters, but I would put the top level module before, e.g.
```ruby
if SETTINGS[:katello][:use_pulp]
  include Katello::Glue
  include Katello::Glue::Pulp::Node
end
```

File: app/models/katello/glue/pulp/node.rb (hunk 11 above)
Comment (ohadlevy) [inline app/models/katello/glue/pulp/node.rb]: is there any outcome? or another pr to look at?

File: app/models/katello/glue/pulp/node.rb [hunk 34]
```diff
@@ -0,0 +1,22 @@
+module Katello
+  module Glue::Pulp::Node
+    def self.included(base)
```
Comment (ohadlevy) [inline app/models/katello/glue/pulp/node.rb]: please switch to using ActiveSupport::Concerns

File: app/views/foreman/smart_proxies/show.html.erb [hunk 35]
```diff
@@ -0,0 +1,7 @@
+<% if @smart_proxy.has_feature?(SmartProxy::PULP_NODE_FEATURE) -%>
+  <div ng-controller="CapsuleContentController">
+    <%= render :file => 'smart_proxies/show' %>
```
Comment (ohadlevy) [inline app/views/foreman/smart_proxies/show.html.erb]: question: why do you need file here?

Comment (ohadlevy) [inline app/views/foreman/smart_proxies/show.html.erb]: hmm, reading this again, I'm not sure I'm following the logic, this means it will render foreman's show vs katello override ?

File: app/models/katello/glue/pulp/node.rb (hunk 11 above)
Comment (tstrachota) [inline app/models/katello/glue/pulp/node.rb]: I discussed and checked with @jlsherrill and we should be good using OAuth credentials. At least for now. There's rework to use certificates coming, but it's independent of this PR.

File: app/models/katello/glue/pulp/node.rb (hunk 34 above)
Comment (jlsherrill) [inline app/models/katello/glue/pulp/node.rb]: I would probably just stick this in smart_proxy_extensions itself, we're slowly moving away from a glue structure of things.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js (hunk 19 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.module.js]: In the end I decided to keep it under capsules-content/ as moving it caused me some additional troubles with loading the modules. I might move it if there's some time left.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js (hunk 21 above)
Comment (tstrachota) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/capsule-content/capsule-content.controller.js]: I moved at least the syncState to a separate service.

File: app/views/foreman/smart_proxies/show.html.erb (hunk 35 above)
Comment (tstrachota) [inline app/views/foreman/smart_proxies/show.html.erb]: Nope, it will always render foreman's show but wrapped with ng-controller for pulp nodes. That's needed to get the angular code working.

**Comment by tstrachota** [comment]
@jlsherrill @ehelms I added tests, fixed your comments and squashed the commits. Please test.

**Comment by ehelms** [comment]
Trying to test this locally, but the content tab doesn't change for me and I see the content view table at the bottom of every tab

**Comment by tstrachota** [comment]
@ehelms this is probably https://github.com/theforeman/foreman/pull/3125

**Comment by ehelms** [comment]
Better now -- the synchronize button isn't making any calls for me and when I try to change pages to a Katello page the URL updates but not the page.

**Comment by tstrachota** [comment]
@ehelms I switched from the fake state to handling the urls in `$urlRouterProvider` manually to fix that issue. The problem was in not having `<ui-view>` element and other proper structure on the page so the state change after click had no effect.

**Comment by tstrachota** [comment]
Rebased.

**Comment by tstrachota** [comment]
There was an issue with compared arrays not being sorted in the tests, should be fixed now.

**Comment by jlsherrill** [comment]
Tested, here's my thoughts:

* Syncing and cancelling worked well :+1: 
* There seem to be two sets of counts, one set on the 'content' tab and another set returned from GET requests to /katello/api/capsules/2-katello-capsule-nightly-example-com/content/sync.  the first set come from the db and I think that makes sense.  The 2nd set are calculated using some complex code which is fine but I don't actually see these being used anywhere.  Should it be visible anywhere on the UI?

**Comment by tstrachota** [comment]
@jlsherrill let me clarify what api calls are used to get the data:

On the page load there's GET `/katello/api/capsules/:capsule_id/content/sync` to retrieve synchronization status. The result is used for:
 - getting sync tasks that are currently running (used to show progress on the first tab)
 - getting last failed sync tasks (used to display progress and errors on the first tab if something is stuck)
 - showing what environments need to be synced (on the first tab)
 - getting counts of synchronized content (used on the Content tab directly)

When there's some task running `AggregateTask` periodically calls GET `/foreman_tasks/api/tasks/bulk_search` to calculate the progress.

So basically you get everything from the GET `.../content/sync` and just load additional task details if there's something running.

**Comment by jlsherrill** [comment]
@tstrachota I figured out my confusion.  You're correct, will continue my reviewing!

File: app/lib/katello/capsule_content.rb [hunk 36]
```diff
@@ -53,6 +53,49 @@ def consumer
...
+    def pulp_repositories_data(environment = nil, content_view = nil)
+      @pulp_repositories ||= @capsule.pulp_repositories
+
+      if environment
+        if content_view
+          filter = /[^-]+-#{environment.label}-#{content_view.label}-[^-]+-[^-]+/
+        else
+          filter = /[^-]+-#{environment.label}-[^-]+-[^-]+-[^-]+/
+        end
+      end
```
Comment (jlsherrill) [inline app/lib/katello/capsule_content.rb]: I think you need to include organization_label here as well.

The other thought and maybe a better way to do it, would be to go through the db and use the repos in the db to filter the list:
```
if environment
  query =   Katello::Repositories.in_environment(environment)
  query = query.in_content_views([content_view]) if content_view
end
pulp_ids = query.pluck(:pulp_id)
@pulp_repositories.select{|r| pulp_ids.include?(r["id"]) }
```

**Comment by jlsherrill** [comment]
@tstrachota just the one comment and I'm good with this.  @ehelms any more concerns from your side?

**Comment by tstrachota** [comment]
@jlsherrill good catch with the organization label! I decided to keep the original filtering method in place. It's not that elegant but it reduces possibility of hiding some repo in case the katello db went out of sync with the proxy.

I also found out that I wasn't scoping the sync status results per organization. Both issues are fixed in my last commit.

File: app/lib/katello/capsule_content.rb [hunk 37]
```diff
@@ -53,6 +53,49 @@ def consumer
...
+
+    def pulp_repositories_data(environment = nil, content_view = nil)
+      @pulp_repositories ||= @capsule.pulp_repositories
+
+      if environment
+        if content_view
+          filter = /#{environment.organization.label}+-#{environment.label}-#{content_view.label}-[^-]+-[^-]+/
+        else
+          filter = /#{environment.organization.label}+-#{environment.label}-[^-]+-[^-]+-[^-]+/
+        end
```
Comment (jlsherrill) [inline app/lib/katello/capsule_content.rb]: one last issue i found is that  the 'default organization view' all show as zero.  I believe the problem is that repos in that content view are different and only have:  org_label-product_label-repo_label.  I still think using the db to figure out labels is better and would handle this case:

```
    def pulp_repositories_data(environment = nil, content_view = nil)
      @pulp_repositories ||= @capsule.pulp_repositories

      repos = Katello::Repository
      repos = repos.in_environment(environment) if environment
      repos = repos.in_content_views([content_view]) if content_view
      puppet_envs = Katello::ContentViewPuppetEnvironment
      puppet_envs = puppet_envs.in_environment(environment) if environment
      puppet_envs = puppet_envs.in_content_view(content_view) if content_view

      repo_ids = repos.pluck(:pulp_id) + puppet_envs.pluck(:pulp_id)

      @pulp_repositories.select{ |r| repo_ids.include?(r['id']) }
    end
```

Comment (jlsherrill) [inline app/lib/katello/capsule_content.rb]: it also doesn't tie the format of the label to anything special, as its not tied to that today even if it is generated with a particular structure

**Comment by jlsherrill** [comment]
Everything else looks okay, i'll merge once that is fixed!

**Comment by tstrachota** [comment]
@jlsherrill You're right, I used your implementation, thanks. Updated and rebased.

**Comment by jlsherrill** [comment]
ACK thanks @tstrachota !
